[JIT] Optimize hot path in ProfilingGraphExecutorImpl::getPlanFor.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47465 [JIT] Optimize hot path in ProfilingGraphExecutorImpl::getPlanFor.**

This results in a ~10% improvement on a DeepAndWide model:

10 runs of a benchmark before the change:
```
1.480785621330142
1.430812582373619
1.3845220785588026
1.4510653037577868
1.4827174227684736
1.3679781593382359
1.4239587392657995
1.5069784726947546
1.3988622818142176
1.4533461946994066
```

10 runs of the same benchmark after the change:
```
1.3221493270248175
1.3624659553170204
1.3415213637053967
1.3560577500611544
1.3064174111932516
1.2934542261064053
1.379274770617485
1.3850531745702028
1.26725466363132
1.3738237638026476
```

Link to benchmark: https://gist.github.com/ZolotukhinM/2308732eabb47685c6f7786e5a13b3d1

Differential Revision: [D24767247](https://our.internmc.facebook.com/intern/diff/D24767247)

Commit:
[JIT] Optimize hot path in ProfilingGraphExecutorImpl::getPlanFor.

This results in a ~10% improvement on a DeepAndWide model:

10 runs of a benchmark before the change:
```
1.480785621330142
1.430812582373619
1.3845220785588026
1.4510653037577868
1.4827174227684736
1.3679781593382359
1.4239587392657995
1.5069784726947546
1.3988622818142176
1.4533461946994066
```

10 runs of the same benchmark after the change:
```
1.3221493270248175
1.3624659553170204
1.3415213637053967
1.3560577500611544
1.3064174111932516
1.2934542261064053
1.379274770617485
1.3850531745702028
1.26725466363132
1.3738237638026476
```

Link to benchmark: https://gist.github.com/ZolotukhinM/2308732eabb47685c6f7786e5a13b3d1

[ghstack-poisoned]

Commit:
Update on "[JIT] Optimize hot path in ProfilingGraphExecutorImpl::getPlanFor."


This results in a ~10% improvement on a DeepAndWide model:

10 runs of a benchmark before the change:
```
1.480785621330142
1.430812582373619
1.3845220785588026
1.4510653037577868
1.4827174227684736
1.3679781593382359
1.4239587392657995
1.5069784726947546
1.3988622818142176
1.4533461946994066
```

10 runs of the same benchmark after the change:
```
1.3221493270248175
1.3624659553170204
1.3415213637053967
1.3560577500611544
1.3064174111932516
1.2934542261064053
1.379274770617485
1.3850531745702028
1.26725466363132
1.3738237638026476
```

Link to benchmark: https://gist.github.com/ZolotukhinM/2308732eabb47685c6f7786e5a13b3d1

Differential Revision: [D24767247](https://our.internmc.facebook.com/intern/diff/D24767247)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -517,6 +511,20 @@ ExecutionPlan ProfilingGraphExecutorImpl::getPlanFor(
+
+  // IMPORTANT: This is a hot path of calling a torchscript function. Try not to
+  // add any code above this.
+  if (optimized_plan_) {
```
I think we still need to check `getGraphExecutorOptimize`

eellison replied:
maybe in debug mode we would also update the optimized_plan_

ZolotukhinM replied:
Yeah, good catch, I'll revert that part. It's a bit of a bummer since `getGraphExecutorOptimize` pokes at TLS and probably was partially responsible for the slowdown, but I agree that even if we're going to change that somehow, it shouldn't be a part of this PR.

eellison reviewed (commented):
Nice speedup! we could consider adding a  version of GRAPH_DUMP that's debug only ?   `GRAPH_DUMP_DEBUG_ONLY` ?

eellison reviewed (approved):
lgtm with the `getGraphExecutorOptimize` check

Commit:
Update on "[JIT] Optimize hot path in ProfilingGraphExecutorImpl::getPlanFor."


This results in a ~10% improvement on a DeepAndWide model:

10 runs of a benchmark before the change:
```
1.480785621330142
1.430812582373619
1.3845220785588026
1.4510653037577868
1.4827174227684736
1.3679781593382359
1.4239587392657995
1.5069784726947546
1.3988622818142176
1.4533461946994066
```

10 runs of the same benchmark after the change:
```
1.3221493270248175
1.3624659553170204
1.3415213637053967
1.3560577500611544
1.3064174111932516
1.2934542261064053
1.379274770617485
1.3850531745702028
1.26725466363132
1.3738237638026476
```

Link to benchmark: https://gist.github.com/ZolotukhinM/2308732eabb47685c6f7786e5a13b3d1

Differential Revision: [D24767247](https://our.internmc.facebook.com/intern/diff/D24767247)

[ghstack-poisoned]
